library/ssl_tls13_server.c

```diff
@@ -1039,8 +1041,13 @@ int ssl_parse_client_psk_identity_ext( mbedtls_ssl_context *ssl,
                                   truncated_clienthello_start, truncated_clienthello_end - truncated_clienthello_start, server_computed_binder );
         }
         else {
-            /* Case 2: We are using an externally configured PSK */
-            ret = ssl_calc_binder( ssl, ssl->conf->psk, ssl->conf->psk_len,
+            /* Case 2: We are using an externally configured PSK, or a dynamically configured PSK if one is defined */
+            if( mbedtls_ssl_get_psk( ssl, &psk, &psk_len ) == MBEDTLS_ERR_SSL_PRIVATE_KEY_REQUIRED )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "should never happen" ) );
+                return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
```

Comment:
This is the wrong error code - `MBEDTLS_ERR_SSL_INTERNAL_ERROR` should only used for assertions, i.e. code that to the best of our knowledge can never be triggered regardless of the user's input. This doesn't seem to be the case here, since we run into this path if the user hasn't configured a PSK. 

Why aren't we returning `MBEDERR_SSL_PRIVATE_KEY_REQUIRED`?

Moreover, why are we ignoring all return values `!= MBEDTLS_ERR_SSL_PRIVATE_KEY_REQUIRED`?

Reply:
Thanks @hanno-arm.
The PSK is never checked after calling `ssl_calc_binder()`, that's why I added this test.
But since the static PSK is checked a few lines above I guess there's no need for that.
I suggest to call `mbedtls_ssl_get_psk()` without checking its return value to maintain status quo.

Reply:
Calling a function which by its spec can fail without acting on the return value is dangerous. Why not just return the error if it fails?

Reply:
Agreed. That's better practice and checking twice is not a problem.